lib/Sema/CSDiagnostics.cpp

```diff
@@ -981,101 +981,133 @@ bool AssignmentFailure::diagnoseAsError() {
   // we find a node in the lvalue path that is problematic, this returns it.
   auto immInfo = resolveImmutableBase(destExpr);
 
-  // Otherwise, we cannot resolve this because the available setter candidates
-  // are all mutating and the base must be mutating.  If we dug out a
-  // problematic decl, we can produce a nice tailored diagnostic.
-  if (auto *VD = dyn_cast_or_null<VarDecl>(immInfo.second)) {
-    std::string message = "'";
-    message += VD->getName().str().str();
-    message += "'";
-
-    auto type = getType(immInfo.first);
-    auto bgt = type ? type->getAs<BoundGenericType>() : nullptr;
-
-    if (bgt && bgt->getDecl() == getASTContext().getKeyPathDecl())
-      message += " is a read-only key path";
-    else if (VD->isCaptureList())
-      message += " is an immutable capture";
-    else if (VD->isImplicit())
-      message += " is immutable";
-    else if (VD->isLet())
-      message += " is a 'let' constant";
-    else if (!VD->isSettable(DC))
-      message += " is a get-only property";
-    else if (!VD->isSetterAccessibleFrom(DC))
-      message += " setter is inaccessible";
-    else {
-      message += " is immutable";
-    }
+  Optional<OverloadChoice> choice = immInfo.second;
+
+  // Attempt diagnostics based on the overload choice.
+  if (choice.hasValue()) {
+
+    auto getKeyPathArgument = [](SubscriptExpr *expr) {
+      auto *TE = dyn_cast<TupleExpr>(expr->getIndex());
+      assert(TE->getNumElements() == 1);
```

Comment:
Nit: Would be great to assert that element has a label `keyPath:` just like original code in `resolveImmutableBase` did.